zio-http/src/main/scala/zio/http/DnsResolver.scala

```diff
@@ -0,0 +1,285 @@
+      patches.foldLeft(entries) { case (entries, patch) =>
+        patch match {
+          case CachePatch.Remove(host)               =>
+            entries - host
+          case CachePatch.Update(host, updatedEntry) =>
+            entries.updated(host, updatedEntry)
+        }
+      }
+  }
+
+  private object CachingResolver {
+    def make(
+      resolver: DnsResolver,
+      ttl: Duration,
+      unknownHostTtl: Duration,
+      maxCount: Int,
+      maxConcurrentResolutions: Int,
+      expireAction: ExpireAction,
+      refreshRate: Duration,
+    ): ZIO[Scope, Nothing, DnsResolver] =
+      for {
+        semaphore <- Semaphore.make(maxConcurrentResolutions)
+        entries   <- Ref.make(Map.empty[String, CacheEntry])
+        cachingResolver = new CachingResolver(resolver, ttl, unknownHostTtl, maxCount, expireAction, semaphore, entries)
+        _ <- cachingResolver.refreshAndCleanup().scheduleFork(Schedule.fixed(refreshRate))
+      } yield cachingResolver
+  }
+
+  private[http] def snapshot(): ZIO[DnsResolver, Nothing, Map[String, CacheEntry]] =
+    ZIO.service[DnsResolver].flatMap {
+      case cachingResolver: CachingResolver => cachingResolver.snapshot()
+      case _ => ZIO.dieMessage(s"Unexpected DnsResolver implementation: ${getClass.getName}")
+    }
+
+  final case class DnsResolverConfig(
+    ttl: Duration,
+    unknownHostTtl: Duration,
+    maxCount: Int,
+    maxConcurrentResolutions: Int,
+    expireAction: ExpireAction,
+    refreshRate: Duration,
+  )
+
+  object DnsResolverConfig {
+    val default: DnsResolverConfig = DnsResolverConfig(
+      ttl = 10.minutes,
+      unknownHostTtl = 1.minute,
+      maxCount = 4096,
+      maxConcurrentResolutions = 16,
+      expireAction = ExpireAction.Refresh,
+      refreshRate = 2.seconds,
+    )
+
+    val config: Config[DnsResolverConfig] =
+      (Config.duration("ttl").withDefault(default.ttl) ++
+        Config.duration("unknown-host-ttl").withDefault(default.unknownHostTtl) ++
+        Config.int("max-count").withDefault(default.maxCount) ++
+        Config.int("max-concurrent-resolutions").withDefault(default.maxConcurrentResolutions) ++
+        ExpireAction.config.nested("expire-action").withDefault(default.expireAction) ++
+        Config.duration("refresh-rate").withDefault(default.refreshRate)).map {
+        case (ttl, unknownHostTtl, maxCount, maxConcurrentResolutions, expireAction, refreshRate) =>
+          DnsResolverConfig(ttl, unknownHostTtl, maxCount, maxConcurrentResolutions, expireAction, refreshRate)
+      }
+  }
+
+  def cached(
+    ttl: Duration = 10.minutes,
+    unknownHostTtl: Duration = 1.minute,
+    maxCount: Int = 4096,
+    maxConcurrentResolutions: Int = 16,
+    expireAction: ExpireAction = ExpireAction.Refresh,
+    refreshRate: Duration = 2.seconds,
+    implementation: DnsResolver = SystemResolver(),
+  ): ZLayer[Any, Nothing, DnsResolver] =
```

Comment:
This is "cached system resolver".

Could be useful to add `cached` as a method on `DnsResolver`.